wrappers/services.go

```diff
@@ -471,116 +471,180 @@ type EnsureSnapServicesOptions struct {
 	RequireMountedSnapdSnap bool
 }
 
-// EnsureSnapServices will ensure that the specified snap services' file states
-// are up to date with the specified options and infos. It will add new services
-// if those units don't already exist, but it does not delete existing service
-// units that are not present in the snap's Info structures.
-// There are two sets of options; there are global options which apply to the
-// entire transaction and to every snap service that is ensured, and options
-// which are per-snap service and specified in the map argument.
-// If any errors are encountered trying to update systemd units, then all
-// changes performed up to that point are rolled back, meaning newly written
-// units are deleted and modified units are attempted to be restored to their
-// previous state.
-// To observe which units were added or modified a
-// ObserveChangeCallback calllback can be provided. The callback is
-// invoked while processing the changes. Because of that it should not
-// produce immediate side-effects, as the changes are in effect only
-// if the function did not return an error.
-// This function is idempotent.
-func EnsureSnapServices(snaps map[*snap.Info]*SnapServiceOptions, opts *EnsureSnapServicesOptions, observeChange ObserveChangeCallback, inter Interacter) (err error) {
-	// note, sysd is not used when preseeding
-	sysd := systemd.New(systemd.SystemMode, inter)
-
-	if opts == nil {
-		opts = &EnsureSnapServicesOptions{}
-	}
-
-	// we only consider the global EnsureSnapServicesOptions to decide if we
-	// are preseeding or not to reduce confusion about which set of options
-	// determines whether we are preseeding or not during the ensure operation
-	preseeding := opts.Preseeding
-
-	// modifiedUnitsPreviousState is the set of units that were modified and the previous
+// ensureSnapServicesContext is the context for EnsureSnapServices.
+// EnsureSnapServices supports transactional update/write of systemd service
+// files and slice files. A part of this is to support rollback of files and
+// also keep track of whether a restart of systemd daemon is required.
+type ensureSnapServicesContext struct {
+	// snaps, observeChange, opts and inter are the arguments
+	// taken by EnsureSnapServices. They are here to allow sub-functions
+	// to easier access these, and keep parameter lists shorter.
+	snaps         map[*snap.Info]*SnapServiceOptions
+	observeChange ObserveChangeCallback
+	opts          *EnsureSnapServicesOptions
+	inter         Interacter
+
+	// note: is not used when preseeding is set in opts.Preseeding
+	sysd systemd.Systemd
+	// modifiedSystemServices/modifiedUserServices keeps track of whether a systemd
+	// daemon reload is necessary, either for the system level and/or user level.
+	modifiedSystemServices bool
+	modifiedUserServices   bool
```

Comment:
nitpick: `systemDaemonReloadNeeded` and `userDaemonReloadNeeded` are IMHO less confusing and probably wouldn't need any explanatory comment at all; the current variable names suggest that these variables contain a list of the modified services (still, your name is better than what we previously had :-) )

Reply:
Good point!